File: frontend/src/types/self-improve.ts
export interface Cycle {
  id: string;
  status:
    | 'idle'
    | 'planning'
    | 'architecting'
    | 'developing'
    | 'testing'
    | 'reviewing'
    | 'documenting'
    | 'awaiting_approval'
    | 'merging'
    | 'integrating'
    | 'completed'
    | 'rejected'
    | 'failed';
  budget_usd: number;
  spent_usd: number;
  max_improvements: number;
  focus_areas: string[] | null;
  improvements: Improvement[];
  agents_spawned: number;
  started_at: string | null;
  completed_at: string | null;
  error: string | null;
}

export interface GateCheck {
  name: string;
  passed: boolean;
  message: string;
  severity: string;
}

export interface GateResult {
  gate: string;
  verdict: 'pass' | 'fail' | 'warn';
  checks: GateCheck[];
  summary: string;
  retry_count: number;
  run_at: string;
  duration_ms: number;
}

export interface Improvement {
  id: string;
  cycle_id: string;
  title: string;
  description: string;
  priority: number;
  category: 'performance' | 'ux' | 'tests' | 'refactor' | 'feature' | 'docs';
  status:
    | 'proposed'
    | 'approved'
    | 'in_progress'
    | 'testing'
    | 'review'
    | 'documenting'
    | 'awaiting_approval'
    | 'approved_by_human'
    | 'rejected_by_human'
    | 'merged'
    | 'failed';
  worktree_path: string | null;
  branch_name: string | null;
  diff_summary: string | null;
  diff_stat: string | null;
  test_results: {
    passed: number;
    failed: number;
    errors: number;
    output: string;
  } | null;
  review_notes: string | null;
  security_scan: {
    passed: boolean;
    issues: string[];
  } | null;
  pr_description: string | null;
  agent_id: string | null;
  gate_results?: GateResult[];
  created_at: string;
  updated_at: string;
}

export interface SquadAgent {
  id: string;
  cycle_id: string;
  improvement_id: string | null;
  agent_id: string;
  role: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
  started_at: string | null;
  completed_at: string | null;
  output_summary: string | null;
}

export interface SquadRole {
  role: string;
  name: string;
  description: string;
  stage: string;
}

export type CycleStatus = Cycle['status'];
export type ImprovementStatus = Improvement['status'];
export type ImprovementCategory = Improvement['category'];

File: backend/app/models/self_improve.py
"""Pydantic models for self-improve endpoints."""

from typing import Optional
from pydantic import BaseModel


class StartCycleBody(BaseModel):
    budget_usd: float = 5.0
    max_improvements: int = 5
    focus_areas: Optional[list[str]] = None  # e.g. ["performance", "ux", "tests"]


class CycleOut(BaseModel):
    id: str
    status: str  # idle, planning, architecting, developing, testing, reviewing, documenting, awaiting_approval, merging, integrating, completed, rejected, failed
    budget_usd: float
    spent_usd: float
    max_improvements: int
    focus_areas: list[str] | None
    improvements: list[dict]  # list of ImprovementOut
    agents_spawned: int
    started_at: str | None
    completed_at: str | None
    error: str | None


class ImprovementOut(BaseModel):
    id: str
    cycle_id: str
    title: str
    description: str
    priority: int  # 1 = highest
    category: str  # performance, ux, tests, refactor, feature, docs
    status: str  # proposed, approved, in_progress, testing, review, documenting, awaiting_approval, approved_by_human, rejected_by_human, merged, failed
    worktree_path: str | None
    branch_name: str | None
    diff_summary: str | None
    diff_stat: str | None  # e.g. "5 files changed, 120 insertions(+), 30 deletions(-)"
    test_results: dict | None  # {passed: int, failed: int, errors: int, output: str}
    review_notes: str | None
    security_scan: dict | None  # {passed: bool, issues: list[str]}
    pr_description: str | None
    agent_id: str | None  # which agent is working on it
    created_at: str
    updated_at: str


class ApproveImprovementBody(BaseModel):
    comment: str | None = None


class RejectImprovementBody(BaseModel):
    reason: str | None = None


class SelfImprovePreferences(BaseModel):
    auto_enabled: bool = False
    cron_expression: str = "0 3 * * 1"  # Default: Weekly Monday 3AM
    max_cost_per_cycle: float = 5.0
    focus_areas: list[str] = []


class CycleAnalyticsItem(BaseModel):
    id: str
    started_at: str | None
    files_changed: int
    cost: float
    duration_seconds: int
    improvements_count: int
    status: str
